File: src/codes/single-state-handling/ts/rfc_command.ts
// コマンドインターフェース
interface PublishCommand {
  execute(): void;
}

// 各状態の処理をコマンド化
class DraftPublishCommand implements PublishCommand {
  execute(): void {
    console.log("レビュー依頼を送信");
  }
}

class ReviewPublishCommand implements PublishCommand {
  execute(): void {
    console.log("公開しました");
  }
}

class PublishedPublishCommand implements PublishCommand {
  execute(): void {
    console.log("すでに公開済みです");
  }
}

// Invoker（呼び出し役）
class Document {
  private command: PublishCommand;

  constructor(command: PublishCommand) {
    this.command = command;
  }

  setCommand(command: PublishCommand): void {
    this.command = command;
  }

  publish(): void {
    this.command.execute();
  }
}

// 利用例
const doc = new Document(new DraftPublishCommand());
doc.publish(); // → レビュー依頼

doc.setCommand(new ReviewPublishCommand());
doc.publish(); // → 公開

doc.setCommand(new PublishedPublishCommand());
doc.publish(); // → すでに公開済み

File: src/codes/single-state-handling/py/rfc_command.py
from abc import ABC, abstractmethod

class PublishCommand(ABC):
    @abstractmethod
    def execute(self):
        pass

class DraftPublishCommand(PublishCommand):
    def execute(self):
        print("レビュー依頼を送信")

class ReviewPublishCommand(PublishCommand):
    def execute(self):
        print("公開しました")

class PublishedPublishCommand(PublishCommand):
    def execute(self):
        print("すでに公開済みです")

class Document:
    def __init__(self, command: PublishCommand):
        self._command = command

    def set_command(self, command: PublishCommand):
        self._command = command

    def publish(self):
        self._command.execute()

# 利用例
doc = Document(DraftPublishCommand())
doc.publish()

doc.set_command(ReviewPublishCommand())
doc.publish()

doc.set_command(PublishedPublishCommand())
doc.publish()
